#include "edge.h"
#include "ui_edge.h"
#include <QVector2D>
#include <QVector3D>
#include <QDebug>
#include <iostream>

using namespace cv;
using namespace std;

Edge::Edge(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::Edge)
{
    ui->setupUi(this);
    setFixedSize(size().width(), size().height());

    // set up threshold slider
    ui->vary_lineEdit->setText("0");
    ui->varySlider->setRange(0, 255);
    ui->varySlider->setSingleStep(1);
    ui->varySlider->setTickPosition(ui->varySlider->TicksBelow);
    ui->varySlider->setTickInterval(5);

} // constructor

Edge::~Edge()
{
    delete ui;
}

void Edge::changeEvent(QEvent *e)
{
    QMainWindow::changeEvent(e);
    switch (e->type()) {
    case QEvent::LanguageChange:
        ui->retranslateUi(this);
        break;
    default:
        break;
    }
}// change window

/*****************************************************************************************************/
/****************************************** EDGE MODE ************************************************/
/*****************************************************************************************************/

void Edge::setImageView(Mat imageIn)
{
    this->src = imageIn;
    Mat imageOut = setEdge(src, 0);
    updateView(imageOut);

} // set image View

Mat Edge::setEdge(Mat imageIn, int thresh_val)
{
    Mat edge;
    Mat imageIn_resize;

    //gray scale
    cv::resize(imageIn, imageIn_resize, cv::Size2i(imageIn.cols/3, imageIn.rows/3));
    cvtColor(imageIn_resize, src_gray, cv::COLOR_BGR2GRAY);
    blur(src_gray, edge, Size(3,3));
    Canny(edge, edge, thresh_val, thresh_val*3, 3);
    cv::resize(edge, edge, cv::Size2i(edge.cols*3, edge.rows*3));
    return edge;

} // edge detection function

void Edge::updateView(Mat imageOut)
{
    /* convert Mat to QImage
     display on graphicsView */
    QImage img((uchar*)imageOut.data, imageOut.cols, imageOut.rows, QImage::Format_Indexed8);
    image = QPixmap::fromImage(img);
    scene = new QGraphicsScene(this);
    scene->addPixmap(image);
    scene->setSceneRect(0, 0, image.width(), image.height());
    ui->graphicsView->setScene(scene);
} // update graphic view


/*****************************************************************************************************/
/****************************************** EDGE WINDOW UI *******************************************/
/*****************************************************************************************************/

void Edge::on_varySlider_valueChanged(int value)
{
    thresh = value;
    ui->vary_lineEdit->setText(QString::number(value));
    Mat imageOut = setEdge(src, value);
    updateView(imageOut);
} // vary slider


/*****************************************************************************************************/
/************************************ BLOOD VESSELS' TIPS DETECTION **********************************/
/*****************************************************************************************************/

Mat Edge::detectTips(Mat imageIn, unordered_map<string, QVector<QVector2D> > &tips_map, string imgName, int thresh_val)
{
    this->src = imageIn;

    // thin the blood vessels
    Mat edge;
    cvtColor(src, src_gray, cv::COLOR_BGR2GRAY);
    blur(src_gray, edge, Size(3,3));
    Canny(edge, edge, thresh_val, thresh_val*3, 3);

    // determine endpoints after thinning
    endp(edge, dst, tips_map, imgName);

    updateView(dst);

    return dst;

}

void Edge::branchGraph(Mat imageIn, Mat &imageOut)
{
    this->src = imageIn;

    // thin the blood vessels
    Mat edge;
    cvtColor(src, src_gray, cv::COLOR_BGR2GRAY);
    blur(src_gray, edge, Size(3,3));
    Canny(edge, edge, 135, 135*3, 3);

    imageOut = edge.clone();
    cv::threshold(imageOut, imageOut, 0, 1, THRESH_BINARY);

    applylut_1(imageOut,imageOut);
    imageOut = imageOut*255;

    skel(edge, imageOut);
}

/*****************************************************************************************************/
/************ The following codes are obtained from stackoverflow.com question #:22058485 ************/
/*****************************************************************************************************/

//-----------------------------------------------------------------------------------------------------
// LUT for skeletonization
//-----------------------------------------------------------------------------------------------------
void Edge::GetLutSkel(Mat& Lut)
{
    Lut=Mat(8,512,CV_16UC1);
    static int lut1[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,0,1,0,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,0,1,0,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
    static int lut2[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
    static int lut3[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,0,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,0,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
    static int lut4[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,1,1,0,0,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,1,1,0,0,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
    static int lut5[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1};
    static int lut6[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,0,0,1,1,0,0,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,0,0,1,1,0,0,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
    static int lut7[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
    static int lut8[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1};

    for (int i=0;i<512;i++)
    {
        Lut.at<unsigned short>(0,i)=lut1[i];
        Lut.at<unsigned short>(1,i)=lut2[i];
        Lut.at<unsigned short>(2,i)=lut3[i];
        Lut.at<unsigned short>(3,i)=lut4[i];
        Lut.at<unsigned short>(4,i)=lut5[i];
        Lut.at<unsigned short>(5,i)=lut6[i];
        Lut.at<unsigned short>(6,i)=lut7[i];
        Lut.at<unsigned short>(7,i)=lut8[i];
    }
}

//-----------------------------------------------------------------------------------------------------
// http://matlab.exponenta.ru/imageprocess/book3/13/applylut.php
//-----------------------------------------------------------------------------------------------------
void Edge::applylut_1(Mat &imageIn,Mat &imageOut)
{
    static int lut_endpoints[] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,0,1,1,1,1,0,1,0,0,0,1,0,0,0,0,
                                  0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,1,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,
                                  0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,1,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,
                                  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,
                                  1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,
                                  1,1,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,1,1,0,0,0,0,0,0,0,1,0,0,0,0,
                                  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,
                                  0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                  1,0,0,0,1,0,1,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,
                                  0,0,0,0,1,1,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,1,1,
                                  1,1,0,1,1,1,1,0};

    Mat k(3,3,CV_16UC1);

    k.at<unsigned short>(0,0)=256;
    k.at<unsigned short>(1,0)=128;
    k.at<unsigned short>(2,0)=64;
    k.at<unsigned short>(0,1)=32;
    k.at<unsigned short>(1,1)=16;
    k.at<unsigned short>(2,1)=8;
    k.at<unsigned short>(0,2)=4;
    k.at<unsigned short>(1,2)=2;
    k.at<unsigned short>(2,2)=1;

    imageOut=imageIn.clone();

        filter2D(imageOut,imageOut,CV_16UC1,k);
        for(int i=0;i<imageOut.rows;i++) {
            for (int j=1;j<imageOut.cols;j++) {
                imageOut.at<unsigned short>(i,j)=lut_endpoints[imageOut.at<unsigned short>(i,j)];
            }
        }

    imageOut.convertTo(imageOut,CV_8UC1);
}

//-----------------------------------------------------------------------------------------------------
// http://matlab.exponenta.ru/imageprocess/book3/13/applylut.php
//-----------------------------------------------------------------------------------------------------
void Edge::applylut_8(Mat &src, Mat &dst, Mat &lut)
{
    Mat k(3, 3, CV_16UC1);

    k.at<unsigned short>(0, 0) = 256;
    k.at<unsigned short>(1, 0) = 128;
    k.at<unsigned short>(2, 0) = 64;
    k.at<unsigned short>(0, 1) = 32;
    k.at<unsigned short>(1, 1) = 16;
    k.at<unsigned short>(2, 1) = 8;
    k.at<unsigned short>(0, 2) = 4;
    k.at<unsigned short>(1, 2) = 2;
    k.at<unsigned short>(2, 2) = 1;

    dst=src.clone();

    for(int I = 7; I >= 0; I--) {
        filter2D(dst, dst, CV_16UC1, k);
        for(int i = 0; i < dst.rows; i++) {
            for (int j = 1; j < dst.cols; j++) {
                dst.at<unsigned short>(i, j) = lut.at<unsigned short>(I, dst.at<unsigned short>(i, j));
            }
        }
    }

    dst.convertTo(dst,CV_8UC1);
}
//-----------------------------------------------------------------------------------------------------
// LUT Skeletonizer
//-----------------------------------------------------------------------------------------------------
void Edge::skel(Mat &src, Mat &dst)
{
    Mat lut;
    GetLutSkel(lut);
    dst = src.clone();

    cv::threshold(dst, dst, 0, 1, THRESH_BINARY);

    int last_pc = INT_MAX;
    for (int pc = countNonZero(dst); pc<last_pc; pc = countNonZero(dst)) {
        last_pc = pc;
        applylut_8(dst, dst, lut);
    }

    dst = dst * 255;
    // imshow("skel", dst); // displays the skeleton of the image in a separate window
}

//-----------------------------------------------------------------------------------------------------
// LUT endpoints
//-----------------------------------------------------------------------------------------------------
void Edge::endp(Mat &imageIn, Mat &imageOut, unordered_map<string, QVector<QVector2D> > &tips_map, string imgName)
{
    imageOut=imageIn.clone();
    cv::threshold(imageOut,imageOut,0,1,THRESH_BINARY);

    applylut_1(imageOut,imageOut);
    imageOut = imageOut*255;

    //imshow("Blood Vessel Tips", imageOut); // display tips in a separate window
    //automated_tips_images.push_back(imageOut);
    getTipsCoords(imageOut, tips_map, imgName);
}

void Edge::getTipsCoords(Mat imageIn, unordered_map<string, QVector<QVector2D> > &tips_map, string imgName)
{
    QVector<QVector2D> pts_temp;
    Mat img = imageIn;
    Mat img_orig_bg = imread(imgName);
    cvtColor(img_orig_bg, img_orig_bg, cv::COLOR_BGR2GRAY);
    cv::resize(img_orig_bg, img_orig_bg, cv::Size2i(imageIn.cols, imageIn.rows));

    QVector2D pt;
    QVector<QVector2D> pts;
    for (int x = 0; x < imageIn.cols/2; x++) {
        for (int y = 0; y < imageIn.rows; y++) {
            int color = imageIn.at<unsigned short>(Point(x,y)); // get pixel color
            if (color != 0) {
                // pt.setX((float)((x*2) - imageIn.cols/2)/(float)(imageIn.cols/2));
                // pt.setY((float)(imageIn.rows/2 - y)/(float)(imageIn.rows/2));

                // new coordinates - units in pixels
                pt.setX((x*2) - imageIn.cols/2);
                pt.setY(imageIn.rows/2 - y);

                pts.push_back(pt);
                pts_temp.push_back(QVector2D(x, y));
            }
        }
    }
    tips_map[imgName] = pts;

    for (int i = 0; i < pts_temp.size(); i++) {
        QVector2D pt = pts_temp.at(i);
        int a = pt.x();
        int b = pt.y();
        Point dot = Point(a*2, b);
        circle(img, dot, 4.0, Scalar(255, 255, 255), -1, 8);
        circle(img_orig_bg, dot, 4.0, Scalar(0, 0, 0), -1, 8);
    }

    automated_tips_images.push_back(img);
    automated_tips_images_origBG.push_back(img_orig_bg);

}

void Edge::convertToPixelCoords(Mat imageIn, unordered_map<string, QVector<QVector2D> > &tips_map, string imgName)
{
    QVector<QVector2D> pts = tips_map[imgName];
    for (int i = 0; i < pts.size(); i++) {
        QVector2D pt = pts.at(i);
        // int x = (pt.x() * imageIn.cols/2) + (imageIn.cols/2);
        // int y = (imageIn.rows/2) - (pt.y() * imageIn.rows/2);

        // new coordinates
        int x = (pt.x() + imageIn.cols/2)/2;
        int y = (imageIn.rows/2 - pt.y());

        pt.setX(x);
        pt.setY(y);
        pts[i] = pt;
    }

    tips_map[imgName] = pts;

}

void Edge::getAutoTipsImages(QVector<Mat> &auto_tips_images)
{
    auto_tips_images = automated_tips_images;
}

void Edge::getAutoTipsImagesWithOrigBG(QVector<Mat> &auto_tips_images_origBG)
{
    auto_tips_images_origBG = automated_tips_images_origBG;
}

void Edge::clearImageVectors()
{
    automated_tips_images.clear();
    automated_tips_images_origBG.clear();
}
